src/Starcounter.Database.Extensions/TransactorBase.cs

```diff
@@ -13,33 +14,70 @@ protected TransactorBase(ITransactor innerTransactor)
-            => _inner.Transact(db => ExecuteCallback(db, function), options);
+        {
+            var context = EnterTransactorContext();
+            var r = _inner.Transact(db => ExecuteCallback(context, db, function), options);
+            LeaveTransactorContext(context);
+            return r;
+        }
 
-        public virtual Task TransactAsync(Action<IDatabaseContext> action, TransactOptions options = null)
-            => _inner.TransactAsync(db => ExecuteCallback(db, action), options);
+        public async virtual Task TransactAsync(Action<IDatabaseContext> action, TransactOptions options = null)
+        {
+            var context = EnterTransactorContext();
+            await _inner.TransactAsync(db => ExecuteCallback(context, db, action), options);
+            LeaveTransactorContext(context);
+        }
 
-        public virtual Task TransactAsync(Func<IDatabaseContext, Task> function, TransactOptions options = null)
-            => _inner.TransactAsync(db => ExecuteCallback(db, function), options);
+        public async virtual Task TransactAsync(Func<IDatabaseContext, Task> function, TransactOptions options = null)
+        {
+            var context = EnterTransactorContext();
+            await _inner.TransactAsync(db => ExecuteCallback(context, db, function), options);
+            LeaveTransactorContext(context);
+        }
 
-        public virtual Task<T> TransactAsync<T>(Func<IDatabaseContext, T> function, TransactOptions options = null)
-            => _inner.TransactAsync(db => ExecuteCallback(db, function), options);
+        public async virtual Task<T> TransactAsync<T>(Func<IDatabaseContext, T> function, TransactOptions options = null)
+        {
+            var context = EnterTransactorContext();
+            var r = await _inner.TransactAsync(db => ExecuteCallback(context, db, function), options);
+            LeaveTransactorContext(context);
+            return r;
+        }
 
-        public virtual Task<T> TransactAsync<T>(Func<IDatabaseContext, Task<T>> function, TransactOptions options = null)
-            => _inner.TransactAsync(db => ExecuteCallback(db, function), options);
+        public async virtual Task<T> TransactAsync<T>(Func<IDatabaseContext, Task<T>> function, TransactOptions options = null)
+        {
+            var context = EnterTransactorContext();
+            var r = await _inner.TransactAsync(db => ExecuteCallback(context, db, function), options);
+            LeaveTransactorContext(context);
+            return r;
+        }
 
         public virtual bool TryTransact(Action<IDatabaseContext> action, TransactOptions options = null)
-            => _inner.TryTransact(db => ExecuteCallback(db, action), options);
+        {
+            var context = EnterTransactorContext();
+            var r = _inner.TryTransact(db => ExecuteCallback(context, db, action), options);
+            
+            if (r)
```

Comment:
Not sure about this condition. And also about LeaveTransactorContext not being called on exceptions. What if we do some bookkeeping in EnterTransactorContext that requires a cleanup.
WTBS, can be fixed later when/if we actually stumble on it.